src/main/kotlin/io/sc3/peripherals/client/item/PosterRenderer.kt

```diff
@@ -0,0 +1,245 @@
+      matrixStack.translate(-64.0f, -64.0f, 0.0f)
+
+      matrixStack.translate(0.0f, 0.0f, -1.0f)
+
+      val k: Int = getLight(itemFrameEntity, 15728850, light)
+
+      draw(
+        matrixStack,
+        vertexConsumerProvider,
+        posterId,
+        posterState,
+        k
+      )
+
+      return true
+    }
+
+    return false
+  }
+
+  fun drawBackground(matrices: MatrixStack, vertexConsumers: VertexConsumerProvider, swingProgress: Int) {
+    val vertexConsumer = vertexConsumers.getBuffer(POSTER_BACKGROUND)
+    val matrix4f = matrices.peek().positionMatrix
+    vertexConsumer.vertex(matrix4f, -7.0f, 135.0f, 0.0f).color(255, 255, 255, 255).texture(0.0f, 1.0f).light(swingProgress).next()
+    vertexConsumer.vertex(matrix4f, 135.0f, 135.0f, 0.0f).color(255, 255, 255, 255).texture(1.0f, 1.0f).light(swingProgress).next()
+    vertexConsumer.vertex(matrix4f, 135.0f, -7.0f, 0.0f).color(255, 255, 255, 255).texture(1.0f, 0.0f).light(swingProgress).next()
+    vertexConsumer.vertex(matrix4f, -7.0f, -7.0f, 0.0f).color(255, 255, 255, 255).texture(0.0f, 0.0f).light(swingProgress).next()
+  }
+
+  fun drawBackgroundCropped(
+    matrices: MatrixStack,
+    vertexConsumers: VertexConsumerProvider,
+    swingProgress: Int,
+    x: Float, y: Float, width: Float, height: Float
+  ) {
+    val vertexConsumer = vertexConsumers.getBuffer(POSTER_BACKGROUND)
+    val matrix4f = matrices.peek().positionMatrix
+    drawCrop(vertexConsumer, matrix4f, swingProgress, 0f, x-7, y-7, width+14, height+14, 128f+14f, 7f, 7f)
+  }
+
+  @Environment(EnvType.CLIENT)
+  internal class PosterTexture(id: String, private var state: PosterState) :
+    AutoCloseable {
+    private val texture: NativeImageBackedTexture = NativeImageBackedTexture(128, 128, true)
+    private val renderLayer: RenderLayer
+    private var needsUpdate = true
+
+    init {
+      val identifier = textureManager.registerDynamicTexture("poster/$id", texture)
+      renderLayer = RenderLayer.getText(identifier)
+    }
+
+    fun setState(state: PosterState) {
+      val bl = this.state !== state
+      this.state = state
+      needsUpdate = needsUpdate or bl
+    }
+
+    fun setNeedsUpdate() {
+      needsUpdate = true
+    }
+
+    fun getRenderColor(colorIndex: Int, brightness: MapColor.Brightness): Int {
+      return if (colorIndex === 0) {
+        0
+      } else {
+        val color = state.palette.getOrNull(colorIndex) ?: return 0
+        val i = brightness.brightness
+        val j: Int = (color shr 16 and 0xFF) * i / 255
+        val k: Int = (color shr 8 and 0xFF) * i / 255
+        val l: Int = (color and 0xFF) * i / 255
+        -0x1000000 or (l shl 16) or (k shl 8) or j
+      }
+    }
+
+    private fun updateTexture() {
+      for (i in 0..127) {
+        for (j in 0..127) {
+          val k = j + i * 128
+          texture.image!!.setColor(j, i, getRenderColor(state.colors[k].toInt(), MapColor.Brightness.NORMAL))
+        }
+      }
```

Comment:
You can avoid the null assertion here by taking texture.image outside of the loop:

```suggestion
      val image = texture.image ?: return
      for (y in 0..127) {
        for (x in 0..127) {
          val i = x + y * 128
          image.setColor(x, y, getRenderColor(state.colors[i].toInt(), MapColor.Brightness.NORMAL))
        }
      }
```

Reply:
You may want to double check my naming of x,y here as I noticed that `allocatePosterId` iterates x-first instead of y-first, have only had a very quick glance